First take at style guide rough draft (#1)
pulumi/frontend-working-group: style-guide.md

# Pulumi Frontend Style Guide

## Intro
At Pulumi, we have a lot of frontend surfaces - our marketing website, docs, Service console, and HubSpot, among others.  This style guide is meant as a collection of recommendations for code across all the surfaces.

The goals of this guide are:
- Create consistency across the many different frontends we have
- Offer a starting point for folks who make frontend changes less frequently
- Lower the amount of discussion needed in PRs about code style decisions

> This is a resource for teams - you might choose to use this style guide fully, or intentionally deviate in some way.  

Reviewer: I love this -- thank you for adding it. ❤️ I care quite a bit about giving people the freedom to make their own choices, and using guides like these to serve more as a helpful reference and reflection of general consensus than as law. ⭐

- https://github.com/airbnb/javascript
- https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/JavaScript
- https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/CSS
- https://google.github.io/styleguide/htmlcssguide.html

## Formatting
### Rely on Prettier whenever possible
Ideally, PRs don’t have to worry about formatting issues (commas, semicolons, spaces, line length).  Prettier can be enabled on auto-save in editors to make consistent formatting effortless.  For the settings you might want to customize, you can use a .prettierrc.json file.
<!-- TODO: Link to example .prettierrc being used in some repos - call out some items like double quotes, 4 spaces -->

### Use ESLint

## File organization + naming conventions
<!-- TODO: Link to Brant’s doc + other relevant resources -->

## Developing for mobile
Consider mobile-first development - build up, don’t pare down.  We use a shared set of breakpoints across our frontend surfaces, and you should consider how your change will appear at each of those breakpoints.  If you intentionally are not considering a breakpoint for any reason, it’s best practice to note that in the code.

> When including different sets of styles for different target viewport sizes using media queries inside the same stylesheet, it is a good idea to make the default styling before any media queries have been applied to the document the narrow screen/mobile styling, and then override this for wider viewports inside successive media queries. [Source](https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/CSS#use_mobile_first_media_queries).

Reviewer: 100% agree, but I had a super hard time parsing this sentence. (And I that see the MDN docs use pretty much the same language.) Maybe something like this? 

  ```suggestion
Write CSS styles for mobile first, then override those styles with media queries targeting successive viewport widths with `@media (min-width: )` ([Source](https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/CSS#use_mobile_first_media_queries)):
```

We might also want to quote our breakpoints here -- but then again, that might be something better suited for more of a design system/living style guide when we have one.

> ```
> /*Default CSS layout for narrow screens*/
> 
> @media (min-width: 480px) {
>   /*CSS for medium width screens*/
> }
> 
> @media (min-width: 800px) {
>   /*CSS for wide screens*/
> }
> 
> @media (min-width: 1100px) {
>   /*CSS for really wide screens*/
> }
> ```

Reviewer: I might add a little example code just to make this a bit clearer: 

```suggestion
/* Default CSS layout for narrow screens. */
button {
    color: black;
}

/* CSS for medium width screens. */
@media (min-width: 480px) {
    button {
        color: red;
    }
}

/* CSS for wide screens. */
@media (min-width: 800px) {
    button {
        color: green;
    }
}

/* CSS for really wide screens. */
@media (min-width: 1100px) {
    button {
        color: blue;
    }
}
```
```

<!-- TODO: link to mobile breakpoint code paths -->

## CSS
Guide to terminology: https://github.com/airbnb/css#terminology

### Naming 

> - Prefer dashes over camelCasing in class names (https://github.com/airbnb/css)

Reviewer: ```suggestion
- Prefer dashes over camelCasing in class names (https://github.com/airbnb/css).
```

- Prefer dash-cased variable names (e.g. $my-variable) over camelCased or snake_cased variable names.
- Use human readable selectors that describe what element(s) they style. [Source](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/).

```
/* Not recommended: meaningless */
.abc-1901 {}

/* Not recommended: presentational */
.button-green {}
.clear {}

/* Recommended: specific */
.gallery {}
.login {}
.video {}

/* Recommended: generic */
.aux {}
.alt {}
```

### Structure

> - Refrain from using over-qualified selectors: `div.container` can simply be stated as `.container`

Reviewer: ```suggestion
- Refrain from using over-qualified selectors: `div.container` can simply be stated as `.container`.
```

- Try to avoid using IDs as selectors, as they are overly specific.

### Property ordering
<!-- TODO: link CSS Wizardry's article on dealing with specificity -->
Group properties together meaningfully.  Here's a suggested ordering from [Wordpress](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/#property-ordering).

1. Display
2. Positioning
3. Box model
4. Colors and Typography
5. @include declarations
6. Nested selectors

### General

> - Write detailed comments for code that isn't self-documenting.  Examples might include: uses of z-index, compatibility or browser-specific hacks
> - Mixins should be used to DRY up your code, add clarity, or abstract complexity--in much the same way as well-named functions.
> - Use flexible/relative units
> - Favor longhand rules over terse shorthand

Reviewer: ```suggestion
- Write detailed comments for code that isn't self-documenting.  Examples might include: uses of z-index, compatibility or browser-specific hacks.
- Mixins should be used to DRY up your code, add clarity, or abstract complexity--in much the same way as well-named functions.
- Use flexible/relative units.
- Favor longhand rules over terse shorthand.
```

> To start with, it is often harder to understand what the shorthand is doing. It takes a while to pick apart exactly what font syntax is doing.  [Source](https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/CSS#favor_longhand_rules_over_terse_shorthand).

Author: ```suggestion
Shorthand is preferable for single properties - for example, `padding: 2px 4px` instead of `padding-top` and `padding-bottom`.  But when combining multiple properties, it is often harder to understand what the shorthand is doing. It takes a while to pick apart exactly what the font syntax below is doing.  [Source](https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Code_guidelines/CSS#favor_longhand_rules_over_terse_shorthand).
```

> `font: small-caps bold 2rem/1.5 sans-serif;`

Reviewer: OMG yes. 😄

It might be good, though, to say that shorthand expressions are okay when they apply to a single property -- like `padding: 2px 4px` and the like, assuming you agree. (As opposed to writing `padding-top`, `padding-right`, etc., everywhere, which would not only be tedious and verbose and easy to mess up, but is also less idiomatic.) What do you think?

Whereas this is more immediate in terms of understanding:
```
font-variant: small-caps;
font-weight: bold;
font-size: 2rem;
line-height: 1.5;
font-family: sans-serif;
```

> ## Javascript

Reviewer: ```suggestion
## JavaScript
```

> 

Reviewer: ```suggestion

### Use TypeScript

We generally prefer TypeScript over vanilla JavaScript. Occasionally this may not be practical, and that's okay, but as a general rule, if you _can_ use TypeScript, you should.

```

### Use property value shorthand.
Why? It is shorter and descriptive. [Source](https://github.com/airbnb/javascript#es6-object-concise).

```
const lukeSkywalker = 'Luke Skywalker';

// bad
const obj = {
  lukeSkywalker: lukeSkywalker,
};

// good
const obj = {
  lukeSkywalker,
};
```

> ### Use array spreads ... to copy arrays.

Reviewer: ```suggestion
### Use array spreads (`...`) to copy arrays.
```

Why? It creates a version of the function that executes in the context of this, which is usually what you want, and is a more concise syntax.
Why not? If you have a fairly complicated function, you might move that logic out into its own named function expression.
[Source](https://github.com/airbnb/javascript#arrows--use-them).

```
// bad
[1, 2, 3].map(function (x) {
  const y = x + 1;
  return x * y;
});

// good
[1, 2, 3].map((x) => {
  const y = x + 1;
  return x * y;
});
```

### Use === and !== over == and !=
[Source](https://github.com/airbnb/javascript#comparison--eqeqeq).

### Use shortcuts for booleans, but explicit comparisons for strings and numbers.
[Source](https://github.com/airbnb/javascript#comparison--shortcuts).

```
// bad
if (isValid === true) {
  // ...
}

// good
if (isValid) {
  // ...
}

// bad
if (name) {
  // ...
}

// good
if (name !== '') {
  // ...
}

// bad
if (collection.length) {
  // ...
}

// good
if (collection.length > 0) {
  // ...
}
```

### Ternaries should not be nested and generally be single line expressions.
[Source](https://github.com/airbnb/javascript#comparison--nested-ternaries).

```
// bad
const foo = maybe1 > maybe2
  ? "bar"

>   : value1 > value2 ? "baz" : null;

Reviewer: :bleeding_eyes:

// split into 2 separated ternary expressions
const maybeNull = value1 > value2 ? 'baz' : null;

// better
const foo = maybe1 > maybe2
  ? 'bar'
  : maybeNull;

// best
const foo = maybe1 > maybe2 ? 'bar' : maybeNull;

> 15.7 Avoid unneeded ternary statements. eslint: no-unneeded-ternary

Reviewer: Should this be ...?

```suggestion
#### Avoid unneeded ternary statements. eslint: no-unneeded-ternary
```

const bar = c ? true : false;
const baz = c ? false : true;

// good
const foo = a || b;
const bar = !!c;
const baz = !c;
```

### TODOs in code comments should always be accompanied by a ticket link
That ticket should ideally include an item to remove the TODO line in the code.

### Use camelCase when naming objects, functions, and instances.
[Source](https://github.com/airbnb/javascript#naming--camelCase).

```
// bad
const OBJEcttsssss = {};
const this_is_my_object = {};
function c() {}

// good
const thisIsMyObject = {};
function thisIsMyFunction() {}
```

### Acronyms and initialisms should always be all uppercased, or all lowercased.
Why? Names are for readability, not to appease a computer algorithm. [Source](https://github.com/airbnb/javascript#naming--Acronyms-and-Initialisms).

```
// bad
import SmsContainer from './containers/SmsContainer';

// bad
const HttpRequests = [
  // ...
];

// good
import SMSContainer from './containers/SMSContainer';

// good 
const HTTPRequests = [ 
  // ... 
] ;

// also good
const httpRequests = [
  // ...
];

// best 
import TextMessageContainer from './containers/TextMessageContainer' ;

// best
const requests = [
  // ...
];
```

> ### You may optionally uppercase a constant only if it (1) is exported, (2) is a const (it can not be reassigned), and (3) the programmer can trust it (and its nested properties) to never change.

Reviewer: I have to say I'm still a fairly strong 👎 on this one. I'd much prefer we didn't use them, as they call outsize attention to themselves, could actually be misleading if used outside of these conventions, and aren't idiomatic enough that their meaning would be clear to someone new to a particular codebase. `const` is both idiomatic and should give us all we need here.

Reviewer 2: lets add this to the agenda for friday's frontend working group meeting.

Reviewer: Sounds good. I've also given the PR the 👍 anyway, as I'm definitely open to whatever the group decides collectively. I'm just one vote. 😄

const bar = c ? true : false;
const baz = c ? false : true;

// good
const foo = a || b;
const bar = !!c;
const baz = !c;
```

### TODOs in code comments should always be accompanied by a ticket link
That ticket should ideally include an item to remove the TODO line in the code.

### Use camelCase when naming objects, functions, and instances.
[Source](https://github.com/airbnb/javascript#naming--camelCase).

```
// bad
const OBJEcttsssss = {};
const this_is_my_object = {};
function c() {}

// good
const thisIsMyObject = {};
function thisIsMyFunction() {}
```

### Acronyms and initialisms should always be all uppercased, or all lowercased.
Why? Names are for readability, not to appease a computer algorithm. [Source](https://github.com/airbnb/javascript#naming--Acronyms-and-Initialisms).

```
// bad
import SmsContainer from './containers/SmsContainer';

// bad
const HttpRequests = [
  // ...
];

// good
import SMSContainer from './containers/SMSContainer';

// good 
const HTTPRequests = [ 
  // ... 
] ;

// also good
const httpRequests = [
  // ...
];

// best 
import TextMessageContainer from './containers/TextMessageContainer' ;

// best
const requests = [
  // ...
];
```

> ### You may optionally uppercase a constant only if it (1) is exported, (2) is a const (it can not be reassigned), and (3) the programmer can trust it (and its nested properties) to never change.
> Why? This is an additional tool to assist in situations where the programmer would be unsure if a variable might ever change. UPPERCASE_VARIABLES are letting the programmer know that they can trust the variable (and its properties) not to change.
> What about all const variables? - This is unnecessary, so uppercasing should not be used for constants within a file. It should be used for exported constants however.
> What about exported objects? - Uppercase at the top level of export (e.g. EXPORTED_OBJECT.key) and maintain that all nested properties do not change. [Source](https://github.com/airbnb/javascript#naming--uppercase).
> 
> ```
> // bad
> const PRIVATE_VARIABLE = 'should not be unnecessarily uppercased within a file';
> 
> // bad
> export const THING_TO_BE_CHANGED = 'should obviously not be uppercased';
> 
> // bad
> export let REASSIGNABLE_VARIABLE = 'do not use let with uppercase variables';
> 
> // ---
> 
> // allowed but does not supply semantic value
> export const apiKey = 'SOMEKEY';
> 
> // better in most cases
> export const API_KEY = 'SOMEKEY';
> 
> // ---
> 
> // bad - unnecessarily uppercases key while adding no semantic value
> export const MAPPING = {
>   KEY: 'value'
> };
> 
> // good
> export const MAPPING = {
>   key: 'value'
> };
> ```
> 

Author: ```suggestion
```
